drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -787,6 +787,120 @@ static int siwx91x_dev_init(const struct device *dev)
 	return 0;
 }
 
+static int convert_z_sl_twt_req_type(enum wifi_twt_setup_cmd z_req_cmd)
+{
+	switch (z_req_cmd) {
+	case WIFI_TWT_SETUP_CMD_REQUEST:
+		return REQUEST_TWT;
+	case WIFI_TWT_SETUP_CMD_SUGGEST:
+		return SUGGEST_TWT;
+	case WIFI_TWT_SETUP_CMD_DEMAND:
+		return DEMAND_TWT;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int siwx91x_set_twt(const struct device *dev, struct wifi_twt_params *params)
+{
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
+	struct siwx91x_dev *sidev = dev->data;
+	int ret;
+
+	__ASSERT(params, "params cannot be a NULL");
+
+	if (FIELD_GET(SIWX91X_INTERFACE_MASK, sl_wifi_get_default_interface()) !=
+		SL_WIFI_CLIENT_INTERFACE) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	if (sidev->state != WIFI_STATE_DISCONNECTED && sidev->state != WIFI_STATE_INACTIVE &&
+	    sidev->state != WIFI_STATE_COMPLETED) {
+		LOG_ERR("Command given in invalid state");
+		return -EBUSY;
+	}
+
+	if (params->negotiation_type != WIFI_TWT_INDIVIDUAL) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	ret = convert_z_sl_twt_req_type(params->setup_cmd);
+	if (ret < 0) {
+		params->fail_reason = WIFI_TWT_FAIL_CMD_EXEC_FAIL;
+		return ret;
+	}
+
+	twt_req.req_type = ret;
+	/*  Device supports Individual TWT only */
+	twt_req.twt_flow_id = params->flow_id;
+	twt_req.negotiation_type = 0;
+
+	if (params->operation == WIFI_TWT_SETUP) {
```

Comment:
I feel `WIFI_TWT_SETUP` and `WIFI_TWT_TEARDOWN` are rather independent. Did you considered to introduce `siwx91x_set_twt_setup()` and  `siwx91x_set_twt_teardown()`? (I assume `twt_req.req_type` is only used with `WIFI_TWT_SETUP`?)